### PR TITLE
ci(NODE-3860): Enable auth on in CI

### PR DESCRIPTION
**Pull request by baileympearson**
### Description

This PR enables auth for tests running in CI and locally.

#### What is changing?

A few key changes

- Auth is enabled by default locally and in CI
- Many tests were failing when auth was enabled - these have been skipped and a follow up ticket has been created
- A new filter, AuthFilter, was created to filter out tests that require or do not require auth to run
- A handful of tests were fixed instead of skipped.  With these tests, it was either easy to identify the problem or it was easier to fix the issue than to skip the tests (sometimes hundreds of tests would fail for the same reason).

##### Is there new documentation needed for these changes?

No.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/tools/runner/index.js
Comment (baileympearson) [inline test/tools/runner/index.js:16]: I'm open to thoughts here.  With this change I wanted to

1. default tests to running with auth when running locally (or in CI)
2. not require changes to the scripts in package.json or require passing in environment variables

If anyone has a better thought about how to accomplish this, I'm open to suggestions

File: test/integration/client-side-encryption/client_side_encryption.spec.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.spec.test.js:42]: We wanted to go back and reassess if these tests would get fixed by the sharded uri fix we had to make

File: test/integration/connection-monitoring-and-pooling/connection.test.js
Comment (nbbeeken) [inline test/integration/connection-monitoring-and-pooling/connection.test.js:91]: Should be an auth disabled requirement instead right?

File: test/integration/connection-monitoring-and-pooling/connection.test.js
Comment (nbbeeken) [inline test/integration/connection-monitoring-and-pooling/connection.test.js:199]: Should be an auth disabled requirement instead right?

File: test/integration/enumerate_databases/enumerate_databases.test.ts
Comment (nbbeeken) [inline test/integration/enumerate_databases/enumerate_databases.test.ts:11]: Let's add MongoDBMetadataUI to the global namespace, and you can `as MongoDBMetadataUI` to get some hinting

File: test/integration/sessions/sessions.test.js
Comment (nbbeeken) [inline test/integration/sessions/sessions.test.js:22]: Can we leave a comment here referencing why we have to do this, and reference NODE-2149

File: test/manual/socks5.test.ts
Comment (nbbeeken) [inline test/manual/socks5.test.ts:21]: These tests are worth taking a gander at why they're an issue, a socks connection will always be used in an auth scenario, I'm confident that the failure we're experiencing here are a miss-config as opposed to a real issue, any chance you have reference to an error message?

File: test/tools/cluster_setup.sh
Comment (nbbeeken) [inline test/tools/cluster_setup.sh:16]: I would be okay with the script always turning on auth unconditionally, if I need to go out of my way to disable it then writing out the command is fine. I'm thinking keeping this wrapper script simple is ideal, and it keeps us all mostly on the same page with local testing. thoughts?

File: test/tools/reporter/mongodb_reporter.js
Comment (nbbeeken) [inline test/tools/reporter/mongodb_reporter.js:232]: super minor, we could edit the type defined at the top of this file instead, but this just convenience hints, no worries.

File: test/tools/runner/index.js
Comment (nbbeeken) [reply test/tools/runner/index.js:16]: It makes sense why you had to do this to make the auth setting globally available, I think we can defer to a better solution in NODE-3714, if you could reference that in a todo here.

File: test/tools/spec-runner/index.js
Comment (nbbeeken) [inline test/tools/spec-runner/index.js:174]: This todo is no longer relvant?

File: test/tools/spec-runner/index.js
Comment (nbbeeken) [inline test/tools/spec-runner/index.js:181]: Why is this also gated within authEnabled? you should be able to run failCommands even when auth is turned on

File: test/tools/spec-runner/context.js
Comment (nbbeeken) [inline test/tools/spec-runner/context.js:54]: Can we import these from utils now?

File: test/tools/unified-spec-runner/unified-utils.ts
Comment (nbbeeken) [inline test/tools/unified-spec-runner/unified-utils.ts:81]: The proceeding TODO is no longer relevant 🙂  yahoo!!

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/tools/cluster_setup.sh
Comment (baileympearson) [reply test/tools/cluster_setup.sh:16]: Yup, I think that's reasonable!

File: test/tools/runner/index.js
Comment (baileympearson) [reply test/tools/runner/index.js:16]: Done!

File: test/tools/reporter/mongodb_reporter.js
Comment (baileympearson) [reply test/tools/reporter/mongodb_reporter.js]: done done!

File: test/manual/socks5.test.ts
Comment (baileympearson) [reply test/manual/socks5.test.ts]: I definitely was seeing failures before but now I don't, for some reason.  Maybe I fixed it somewhere else?  Regardless, I re-enabled them and ran a patch [here](https://spruce.mongodb.com/version/61facc97d6d80a4fb31284ad/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) and they're passing so I reverted this change.

File: test/integration/connection-monitoring-and-pooling/connection.test.js
Comment (baileympearson) [reply test/integration/connection-monitoring-and-pooling/connection.test.js]: yes good catch!

File: test/integration/client-side-encryption/client_side_encryption.spec.test.js
Comment (baileympearson) [reply test/integration/client-side-encryption/client_side_encryption.spec.test.js:42]: Screen Shot 2022-02-02 at 2 37 04 PM

Our fix to the test runner didn't fix these tests.  These tests seem to be failing with an "extra ping command".  I can dig into these and see if I can find a fix but I'd rather wait and fix them as a part of the follow up ticket

Comment (nbbeeken) [reply test/integration/client-side-encryption/client_side_encryption.spec.test.js:42]: Ah ok yea at least we confirmed its not about the content of the test, I'm fine with defering

**Review by nbbeeken — APPROVED** [review]

File: test/tools/runner/filters/auth_filter.js
Comment (dariakp) [inline test/tools/runner/filters/auth_filter.js:32]: could we throw instead so that we catch typos/unsupported values?

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/integration/enumerate_databases/enumerate_databases.test.ts
Comment (nbbeeken) [reply test/integration/enumerate_databases/enumerate_databases.test.ts:9]: +1 to Daria's comment I think we can remove the limiting altogether

But if you don't mind looks like I didn't get the types quite right for the topology, you could throw this in global.d.ts to improve the hinting  
```ts
import { OneOrMore } from './src/mongo_types';
type TopologyTypeRequirement = OneOrMore<TopologyTypes> | OneOrMore<WithExclusion<TopologyTypes>>
```

File: test/integration/node-specific/mongo_client.test.js
Comment (nbbeeken) [reply test/integration/node-specific/mongo_client.test.js:40]: The raw option has issues when passed to the MongoClient, enabling auth revealed it, but this test also doesn't need to use it, so there's not really a TODO to leave behind. related: NODE-3946

Comment (dariakp) [reply test/integration/node-specific/mongo_client.test.js:40]: Ok, sure

File: test/integration/connection-monitoring-and-pooling/connection.test.js
Comment (baileympearson) [reply test/integration/connection-monitoring-and-pooling/connection.test.js:92]: Yeah, it's skipped right now because it's failing when auth is enabled.  Would you want me to handle it differently?

File: test/tools/spec-runner/utils.js
Comment (baileympearson) [reply test/tools/spec-runner/utils.js:32]: Because some tests specifically do pass in a value for `authSource` and if that value is not `admin`, we don't want to override it.  We just want to default to `admin`

Comment (dariakp) [reply test/tools/spec-runner/utils.js:32]: is that line equivalent to `context?.authSource || 'admin'`?

File: test/tools/reporter/mongodb_reporter.js
Comment (baileympearson) [reply test/tools/reporter/mongodb_reporter.js:236]: I added this specifically for the deadlock tests, which all share a metadata object but all needed to be skipped.  Rather than adding the skipReason for each test individually, I skipped the whole suite by attaching it to the metadata. 

I actually prefer the metadata approach because it keeps all data related to why a test may be skipped together but don't feel strongly about this.  If you'd prefer to have a single method for attaching skip reasons to tests, I can revert this change

File: test/tools/spec-runner/utils.js
Comment (baileympearson) [reply test/tools/spec-runner/utils.js:32]: Yeah, you could also use an `||`.  I can make that change.  

Unfortunately we can't use optional chaining in our raw `.js` tests because we still test on Node v12 and optional chaining wasn't added until v13

File: test/tools/spec-runner/index.js
Comment (baileympearson) [reply test/tools/spec-runner/index.js:159]: I think this is actually unnecessary, since `isAnyRequirementSatisfied` already checks for `auth` requirements.

File: .evergreen/config.yml
Comment (nbbeeken) [reply .evergreen/config.yml:2308]: Slack convo capture: we want to expand the noauth variants to one of each topology, only 5.0 version

File: test/integration/connection-monitoring-and-pooling/connection.test.js
Comment (nbbeeken) [reply test/integration/connection-monitoring-and-pooling/connection.test.js:92]: I'm resolving the comments pointing out tests that would get skipped, our slack convo covered what topologies we want to see run w/ noauth.

File: test/integration/transactions/transactions.spec.test.js
Comment (dariakp) [reply test/integration/transactions/transactions.spec.test.js:100]: I'm not sure I understand the comment regarding "with auth is enabled" - do you mean it's failing on both auth AND noauth? If it's only failing on one or the other, we should still let it run on whichever configuration option is supported by adding it conditionally

File: test/tools/spec-runner/utils.js
Comment (dariakp) [reply test/tools/spec-runner/utils.js]: ah yes, good call out on the chaining

File: test/integration/transactions/transactions.spec.test.js
Comment (baileympearson) [reply test/integration/transactions/transactions.spec.test.js:100]: Yes, it fails when auth is enabled and when it is disabled

File: test/integration/transactions/transactions.spec.test.js
Comment (dariakp) [inline test/integration/transactions/transactions.spec.test.js:96]:
```suggestion
  //    Unlike most of the failing auth tests, this is failing with and without auth enabled.
```

File: test/tools/spec-runner/index.js
Comment (dariakp) [inline test/tools/spec-runner/index.js:393]: we probably want to reverse this order, since that's what is making the spec tests here fail; does reversing the order back to what it was cause other tests to fail?

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/tools/reporter/mongodb_reporter.js
Comment (dariakp) [reply test/tools/reporter/mongodb_reporter.js:236]: I think having multiple ways to use a feature makes that feature more brittle (i.e., higher chance of breaking without noticing in one place or the other), so my personal preference would be to avoid it; but I appreciate the convenience and the visibility afforded by having it in metadata, so if you don't find the risk of the brittleness compelling here, I'm fine with leaving it as is

File: test/manual/tls_support.test.js
Comment (dariakp) [reply test/manual/tls_support.test.js]: Did we confirm that this test runs fine in CI now?

File: test/tools/reporter/mongodb_reporter.js
Comment (nbbeeken) [reply test/tools/reporter/mongodb_reporter.js:236]: What about keeping the `skipReason` made by the metadata filters and appending additional context. "skipped because of an auth disabled requirement" is still useful to see, but then we could have a `skipContext` or `metaSkipReason` (or maybe keep it simple `note: 'xxx'`) that if defined appends more to the message so that the logs reflect both, skipped cus auth and cus of a TODO.

Comment (dariakp) [reply test/tools/reporter/mongodb_reporter.js:236]: The funny thing about this logic is that it's going to append that reason right now no matter what the cause of the skip is, so, for example, when they don't run on single topologies, we'd still get the same metadata skipReason, if I am reading this correctly, since the skipReason isn't attached to a specific filter

Comment (nbbeeken) [reply test/tools/reporter/mongodb_reporter.js:236]: Okay I think that's a valid point and that could lead to confusion, we have the skipReason telling us the reason was auth and we'll keep the TODO to track the unskipping for the particular environment with no auth enabled. I'm okay with that.

**Review by nbbeeken — APPROVED** [review]

**Review by durran — APPROVED** [review]

**Review by dariakp — APPROVED** [review]
This is fine to merge, as long as we have a follow up PR on the same ticket to clean up the skipReason logic

**Pull request merged**
